cmd/if-mover/main.go

```diff
@@ -0,0 +1,185 @@
+	if err != nil {
+		log.Fatal(err)
+	}
+	hostNsName := string(hostNsBytes)
+
+	// Original namespace is kept for switching back
+	log.Println("Getting namespace handles")
+
+	origNs, err := netns.Get()
+	if err != nil {
+		log.Fatal(err)
+	}
+	defer origNs.Close()
+
+	hostNs, err := netns.GetFromName(hostNsName)
+	if err != nil {
+		log.Fatal(err)
+	}
+	defer hostNs.Close()
+
+	log.Println("Switching to host namespace to determine the new interface name")
+	// Note from github.com/vishvananda/netlink documentation:
+	// "
+	// After locking a goroutine to its current OS thread with
+	// runtime.LockOSThread() and changing its network namespace,
+	// any new subsequent goroutine won't be scheduled on that
+	// thread while it's locked. Therefore, the new goroutine will
+	// run in a different namespace leading to unexpected results.
+	// "
+	// This won't be a problem here as long as the program stays as 1 goroutine.
+	// Otherwise this would be needed:
+	// runtime.LockOSThread()
+	// defer runtime.UnlockOSThread()
+
+	if err := netns.Set(hostNs); err != nil {
+		log.Fatal(err)
+		return
+	}
+
+	log.Println("Finding the most recent multus network link on the host namespace")
+	interfaces, err := net.Interfaces()
+	if err != nil {
+		log.Fatal(err)
+		return
+	}
+
+	linkNumber := -1
+	for _, iface := range interfaces {
+		if idStrs := strings.Split(iface.Name, ifBase); len(idStrs) > 1 {
+			id, err := strconv.Atoi(idStrs[1])
+			if err != nil {
+				log.Fatal(err)
+				return
+			}
+			if id > linkNumber {
+				linkNumber = id
+			}
+		}
+	}
+	linkNumber += 1
+
+	newLinkName := fmt.Sprintf("%s%d", ifBase, linkNumber)
+
+	log.Println("Switching back to the original namespace to migrate the link")
+	if err := netns.Set(origNs); err != nil {
+		log.Fatal(err)
+		return
+	}
+
+	log.Println("Finding the multus network connected link")
+	interfaces, err = net.Interfaces()
+	if err != nil {
+		log.Fatal(err)
+		return
+	}
+
+	var link netlink.Link
+
+	for _, iface := range interfaces {
+		if iface.Name == multusLink {
+			link, err = netlink.LinkByName(iface.Name)
+			if err != nil {
+				log.Fatal(err)
+				return
+			}
+			break
+		}
+	}
+
+	if link == nil {
+		log.Fatal("Link not found")
+		return
+	}
+
+	log.Println("Setting multus link down to be renamed")
+	if err := netlink.LinkSetDown(link); err != nil {
+		log.Fatal(err)
+		return
+	}
+
+	log.Printf("Renaming multus link to %s", newLinkName)
+	if err := netlink.LinkSetName(link, newLinkName); err != nil {
+		log.Fatal(err)
+		return
+	}
+
+	// After renaming the link, the link object must be updated or netlink will get confused.
+	link, err = netlink.LinkByName(newLinkName)
+	if err != nil {
+		log.Fatal(err)
+		return
+	}
+
+	log.Println("Determining the IP address of the multus link")
+	addrs, err := netlink.AddrList(link, 0)
+	if err != nil {
+		log.Fatal(err)
+		return
+	}
+
+	// Assuming that the needed address is first on the list.
+	mAddr := addrs[0]
```

Comment:
What do we need the address for?

Reply:
When the interface is moved to the host network namespace, the IP address isn't carried with it, so the interface needs to be reconfigured after it has been moved. I will add this as a comment in the code.